meter/rct.go

```diff
@@ -0,0 +1,156 @@
+package meter
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+
+	"github.com/mlnoga/rct"
+)
+
+/*
+This meter supports devices implementing the RCT communication protocol, e.g. the RCT PS 6.0 with / without battery.
+
+** Usages **
+The following usages are supported:
+- grid    ... for reading the power imported or exported to the grid
+- pv      ... for reading the power produced by the pv
+- battery ... for reading the power imported or exported to the battery
+
+** Example configuration **
+meters:
+- name: GridMeter
+  type: rct
+  host: 192.168.1.23
+  cache: 2s
+  usage: grid
+- name: PvMeter
+  type: rct
+  host: 192.168.1.23
+  cache: 2s
+  usage: pv
+- name: BatteryMeter
+  type: rct
+  host: 192.168.1.23
+  cache: 2s
+  usage: battery
+
+** Limitations **
+It is not allowed to provide different URIs or passwords for different lgess meters since always the
+same hardware instance is accessed with the different usages.
+*/
+
+// RCTMeter implements the api.Meter interface
+type RCTMeter struct {
+	conn    *rct.Connection // connection with the RCT device
+	usage    string         // grid, pv, battery
+}
+
+func init() {
+	registry.Add("rct", NewRCTFromConfig)
+}
+
+//go:generate go run ../cmd/tools/decorate.go -f decorateRCT -b api.Meter -t "api.MeterEnergy,TotalEnergy,func() (float64, error)" -t "api.Battery,SoC,func() (float64, error)"
```

Comment:
The -b base type MUST be `*RCTMeter` meter here or it would swallow the optional interfaces when decorating. Unfortunately this kind of error is impossible to catch without unit tests.

Reply:
A, that helps. Then lgess probably has the same issue, I copied from there. Let me try this. 



Ok, that helps, let me try. 

Not quite sure what the generation is needed for, though. Also, why is generated code checked into the repo, instead of generating as part of the build?

Reply:
> Not quite sure what the generation is needed for, though.

It's exposing optional behaviour. May not strictly be needed here but then it's good for consistency.

> Also, why is generated code checked into the repo, instead of generating as part of the build?

User experience with compiling or making varies :)